apps/api/src/app/workflows-v2/usecases/create-variables-object/create-variables-object.usecase.ts

```diff
@@ -0,0 +1,121 @@
+import { isStringifiedMailyJSONContent } from '../../../environments-v1/usecases/output-renderers/maily-to-liquid/wrap-maily-in-liquid.command';
+import { MailyAttrsEnum } from '../../../environments-v1/usecases/output-renderers/maily-to-liquid/maily.types';
+
+@Injectable()
+export class CreateVariablesObject {
+  constructor(private readonly controlValuesRepository: ControlValuesRepository) {}
+
+  @InstrumentUsecase()
+  async execute(command: CreateVariablesObjectCommand): Promise<Record<string, unknown>> {
+    const controlValues = await this.getControlValues(command);
+
+    const variables = this.extractAllVariables(controlValues);
+    const arrayVariables = [
+      ...this.extractMailyAttribute(controlValues, MailyAttrsEnum.EACH_KEY),
+      ...this.extractMailyAttribute(controlValues, MailyAttrsEnum.ID, this.extractArrayPath),
+    ];
+    const showIfVariables = this.extractMailyAttribute(controlValues, MailyAttrsEnum.SHOW_IF_KEY);
+
+    return keysToObject(variables, arrayVariables, showIfVariables);
+  }
+
+  private async getControlValues(command: CreateVariablesObjectCommand) {
+    if (command.controlValues) {
+      return [command.controlValues].flat().flatMap((obj) => Object.values(obj));
+    }
+
+    if (command.workflowId) {
+      const controls = await this.controlValuesRepository.find(
+        {
+          _environmentId: command.environmentId,
+          _organizationId: command.organizationId,
+          _workflowId: command.workflowId,
+          level: ControlValuesLevelEnum.STEP_CONTROLS,
+          controls: { $ne: null },
+        },
+        {
+          controls: 1,
+          _id: 0,
+        }
+      );
+
+      return controls
+        .map((item) => item.controls)
+        .flat()
+        .flatMap((obj) => Object.values(obj));
+    }
+
+    return [];
+  }
+
+  /**
+   * Extracts all variables from control values by parsing handlebars syntax {{variable}}.
+   * Removes duplicates from the final result.
+   *
+   * @example
+   * controlValues = [ "John {{name}}", "Address {{address}} {{address}}", "nothing", 123, true ]
+   * returns = [ "name", "address" ]
+   */
+  @Instrument()
+  private extractAllVariables(controlValues: unknown[]): string[] {
+    const variables = controlValues.flatMap((value) => {
+      const templateVariables = buildVariables(undefined, value);
+
+      return templateVariables.validVariables.map((variable) => variable.name);
+    });
+
+    return [...new Set(variables)];
+  }
+
+  /**
+   * Extracts variables from Maily JSON content by looking for attribute patterns.
+   * Can optionally transform the extracted values.
+   *
+   * @example
+   * For EACH_KEY: '{"each": "payload.comments"}' returns ["payload.comments"]
+   * For ID with array transform: '{"id": "payload.foo[0].bar"}' returns ["payload.foo"]
+   * For SHOW_IF_KEY: '{"showIfKey": "payload.isHidden"}' returns ["payload.isHidden"]
+   */
+  private extractMailyAttribute(
```

Comment:
I understand that using regex on structured JSON content is generally considered an anti-pattern:

But to give rationale behind the regex approach:
1. Avoids complexity of traversing nested Maily document structure
2. Simple pattern matching is sufficient for our use case (extracting specific string attributes)
3. More performant than parsing + traversing entire document tree for single attribute lookup

In this specific case, since we do various transformations, combining Maily + Liquid + some custom logic, the simplicity outweigh the downsides of a "proper" document traversal solution. The regex pattern is focused and maintainable, targeting only the exact string patterns we need.